prog/dftb+/lib_dftbplus/specieslist.F90

```diff
@@ -0,0 +1,90 @@
+!--------------------------------------------------------------------------------------------------!
+!  DFTB+: general package for performing fast atomistic simulations                                !
+!  Copyright (C) 2006 - 2020  DFTB+ developers group                                               !
+!                                                                                                  !
+!  See the LICENSE file for terms of usage and distribution.                                       !
+!--------------------------------------------------------------------------------------------------!
+
+#:include 'common.fypp'
+
+!> Routines to deal with HSD species lists
+module dftbp_specieslist
+  use dftbp_accuracy, only : dp
+  use dftbp_constants, only : pse
+  use dftbp_hsdutils, only : getChildValue, getChild
+  use dftbp_xmlf90, only : fnode
+  implicit none
+  private
+
+  public :: readSpeciesList
+
+
+  !> Generic wrapper for species specific data
+  interface readSpeciesList
+    module procedure :: readSpeciesListReal
+  end interface readSpeciesList
+
+
+contains
+
+
+  !> Read a list of real valued species data
+  subroutine readSpeciesListReal(node, speciesNames, array, default, conv)
+
+    !> Node to process
+    type(fnode), pointer :: node
+
+    !> Data array to read
+    real(dp), intent(out) :: array(:)
+
+    !> Names of all species
+    character(len=*), intent(in) :: speciesNames(:)
+
+    !> Data array to read
+    real(dp), intent(in), optional :: default(:)
+
+    !> Conversion factor
+    real(dp), intent(in), optional :: conv
+
+    type(fnode), pointer :: value1, child
+    real(dp) :: fact, dummy
+    integer :: iSp
+
+    if (present(conv)) then
+      fact = 1.0_dp / conv
+    else
+      fact = 1.0_dp
+    end if
+
+    if (present(default)) then
+      do iSp = 1, size(speciesNames)
+        call getChild(node, speciesNames(iSp), value1, requested=.false.)
+        if (associated(value1)) then
+          call getChildValue(node, speciesNames(iSp), array(iSp), &
+              & child=child)
+        else
+          call getChildValue(node, speciesNames(iSp), array(iSp), &
+              & default=default(iSp)*fact, child=child)
+        end if
+      end do
+    else
+      do iSp = 1, size(speciesNames)
+        call getChildValue(node, speciesNames(iSp), array(iSp), child=child)
+      end do
+    end if
+
+    do iSp = 1, size(pse)
+      call getChild(node, pse(iSp), value1, requested=.false.)
+      if (associated(value1)) then
```

Comment:
The `getChild()` call already sets the processed flag, you do not need the `if` block afterwards.

Reply:
But we should verify the node could be read into the respective type, otherwise the processed input might contain a string value instead of real there.

Reply:
Removing the if leads to accepting input like
```cson
    Descreening = Values {
      H  =    0.59677857
      C  =    0.87412731
      O  =    0.94134128
      Se = "real number"
    }
```

Reply:
:+1 But then at least, we could spare the repeated search in the linked lists of children by re-using `value` as `getChildValue(value1, "", ...)` within the `if` branch.

Reply:
(By the way, could we then call `value1` maybe `child` instead to make the intention more explicit?) :wink: